java/arcs/android/storage/service/BindingContext.kt

```diff
@@ -112,13 +113,16 @@ class BindingContext(
     ) {
         bindingContextStatisticsSink.traceTransaction("sendProxyMessage") {
             bindingContextStatisticsSink.measure(coroutineContext) {
-                resultCallback.takeIf { it.asBinder().isBinderAlive }?.onResult(null)
-
                 val activeStore = store.activate() as ActiveStore<CrdtData, CrdtOperation, Any?>
                 val actualMessage = message.actual as ProxyMessage<CrdtData, CrdtOperation, Any?>
 
                 if (activeStore.onProxyMessage(actualMessage)) {
+                    resultCallback.takeIf { it.asBinder().isBinderAlive }?.onResult(null)
```

Comment:
this will increase your latency. I don't think it is needed. If the op does not apply the store will request a model sync.

Reply:
I don't think so.
There are at least two cases where the op does not apply at the storage server side:
a) fail to apply / merge the op
b) server crash

The both case will send a success back to the sender w/o second thought who in turn calls its onUpdate() callback immediately. Is this what we want on client side?!!

Reply:
Stability always comes first and never be a compromise.

Reply:
The client's storage proxy would never request a synchronization in such a case as the result is acknowledged as a success. What the proxy would do is call the registered handle callbacks.

Reply:
I think a) is already handled: if the store gets a message and cannot apply it, it will request a sync (it is fine for the proxy to ack with the handles, the write will be committed).
So we only need to fix b). I am not sure how to fix it (maybe the server can be restarted by the ServiceStore?)

Reply:
For a), before the store requests a resync, the success has already been acknowledged back to the client who may be running the onUpdate() callback and kicking off new data logic.

Is this really intended? I really feel this is weird w.r.t onUpdate() gets called with potentially-never-committed data. The correct logic should be the server requests a sync to the client to ensure they have the synced data view, then the client bases on the synced data to generate a new op / model changes. The old op could be obsolete and never be applied successfully due to non-synced data view.

Reply:
adding Shane and Myk to comment on this, especially on desired behavior for a)

Reply:
Yes, this is intended. The whole point of CRDTs is to ensure local writes will always work, and eventual consistency will always be achieved. There is no single authority on the data - every node in the CRDT domain is a source of truth, so there is no concept of data not being committed (at least at runtime).

In our particular architecture, we have a series of "upstream fan-outs" within a given CRDT domain:
- A particle writes to a handle, which in turn writes to the storage proxy. The proxy then notifies any other handles that are connected to it. This is (notionally, at least) occurring within the confines of a single isolation container.
- The proxy writes to its store object (i.e. the data is passed out of the isolation container). The store then notifies any other proxies in other containers. This is acting within a single arc host.
- The store writes to its backing service (e.g. ramdisk, firebase, sql db). The service then notifies any other stores observing the same data. This now allows propagation to any other arcs system (multi-device, cloud services, etc).

We might consider data that is intended to persist after a given arc has shut down as only committed when it reaches its persistent storage space, but even then that space is only the authority on startup. Once other CRDT nodes have been established, they have equal dominion on what a given model holds.

Reply:
I knew how basic operation-based crdt works. I'm confused about how the last step (The store writes to its backing service) fulfills crdt contracts by arcs crdt design on Android when the backing service cannot apply the operation. This will happen at several real cases I can imagine on Android i.e. the ipc/binder shared buffer size is hardcoded 1MB at both storage service and client side respectively, so once too many clients send data or too many big chunks of data are sent to the server side, the ipc/binder driver will raise an exception (not kill server/client) barking out-of-ipc-memory, etc.

Say there is a Singleton storage (ramdisk, db, etc), two handles (A and B) at two different arc hosts writing to the Singleton. A sent the model update a little earlier than B did then unfortunately B's update failed to apply due to out-of-memory or other reasons. B didn't know this at all so moves on at its onUpdate() which changed the model again and the new update was also merged in B's local proxy successfully. In this case, B's first update is never observed and commutated to the storage service and A.

IIUC operation-based crdt works under few assumptions (pls correct me if i'm wrong). At least all locally-successfully-merged operations need to be commutated to any other replicas who are observing model changes. If the assumption may not be able to be fulfilled on some ports like Android port, we need to seek a way-out of how to handle such surprises breaking crdt contracts.

Reply:
When something like this happens (operations dropped or whatnot), we can rely on a future operation failing to apply and a Sync<->ModelUpdate happening between the storage service and the client... It would be good to have some sort of mechanism for the storage proxy to be able to hear of an ack, and if a timeout is missed for some reason - initiate a sync itself to try-again.  This will be handled with my idea in b/156765489, i hope.